### PR TITLE
Fixes #3236 - Adding Maven Central fallback for gradle

### LINKED ISSUE #3236: Dependabot scanning does not work with Gradle  if private only repos defined
**Package manager/ecosystem**

Gradle

**Manifest contents prior to update**

Link to build.gradle 
https://github.com/zendern/private-repo-dependabot-fail/blob/f75be4f5414c17c5194788ab0cf59a26c61bed95/build.gradle#L18

Link to settings.gradle
https://github.com/zendern/private-repo-dependabot-fail/blob/f75be4f5414c17c5194788ab0cf59a26c61bed95/settings.gradle#L3

**Updated dependency**

org.springframework.boot
2.3.9.RELEASE to 2.4.3

org.testcontainers:postgresql
1.15.1 to 1.15.2

org.projectlombok:lombok
1.18.12 to 1.18.18

See here 
https://github.com/zendern/private-repo-dependabot-fail/pull/1

This was opened before i only enabled the private repo and removed maven central. 
Commit to do that was here. It came about a minute after that pr 🤦 
https://github.com/zendern/private-repo-dependabot-fail/commit/f75be4f5414c17c5194788ab0cf59a26c61bed95

**What you expected to see, versus what you actually saw**

No prs are opened. All calls 404. See logs from Insights -> Dependency Graph -> Dependabot scan results. 

```
proxy | time="2021-03-05T02:41:05Z" level=info msg="proxy starting" commit=f762c96c6bc2b9e201fee22268603572200fa968
  proxy | 2021/03/05 02:41:05 Listening (:1080)
updater | 2021-03-05T02:41:05.670503792 [91984925:main:WARN:src/devices/src/legacy/serial.rs:434] Detached the serial input due to peer close/error.
updater | time="2021-03-05T02:41:07Z" level=info msg="guest starting" commit=b387dc9b622a59fd6374cdf69144f884a8af2a85
updater | time="2021-03-05T02:41:07Z" level=info msg="starting job..." fetcher_timeout=5m0s job_id=91984925 updater_timeout=45m0s updater_version=0.134.0-eda24cc2d78adc3babad067e61134f457ea0a6bd
updater | I, [2021-03-05T02:41:08.875869 #7]  INFO -- sentry: ** [Raven] Raven 3.1.1 ready to catch errors
updater | INFO <job_91984925> Starting job processing
  proxy | 2021/03/05 02:41:11 [002] GET https://api.github.com:443/repos/zendern/private-repo-dependabot-fail
  proxy | 2021/03/05 02:41:11 [002] * authenticating github api request
  proxy | 2021/03/05 02:41:11 [002] 200 https://api.github.com:443/repos/zendern/private-repo-dependabot-fail
  proxy | 2021/03/05 02:41:11 [004] GET https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/git/refs/heads/main
  proxy | 2021/03/05 02:41:11 [004] * authenticating github api request
  proxy | 2021/03/05 02:41:11 [004] 200 https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/git/refs/heads/main
  proxy | 2021/03/05 02:41:11 [006] GET https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
  proxy | 2021/03/05 02:41:11 [006] * authenticating github api request
  proxy | 2021/03/05 02:41:11 [006] 200 https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
  proxy | 2021/03/05 02:41:11 [008] GET https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/build.gradle?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
  proxy | 2021/03/05 02:41:11 [008] * authenticating github api request
  proxy | 2021/03/05 02:41:11 [008] 200 https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/build.gradle?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
  proxy | 2021/03/05 02:41:11 [010] GET https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/settings.gradle?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
  proxy | 2021/03/05 02:41:11 [010] * authenticating github api request
  proxy | 2021/03/05 02:41:11 [010] 200 https://api.github.com:443/repos/zendern/private-repo-dependabot-fail/contents/settings.gradle?ref=8b5ab1705717aa59057dd98b98f6d667f0923f0e
updater | INFO <job_91984925> Finished job processing
updater | time="2021-03-05T02:41:11Z" level=info msg="task complete" container_id=job-91984925-file-fetcher exit_code=0 job_id=91984925 step=fetcher
updater | I, [2021-03-05T02:41:12.838025 #8]  INFO -- sentry: ** [Raven] Raven 3.1.1 ready to catch errors
updater | INFO <job_91984925> Starting job processing
updater | INFO <job_91984925> Starting update job for zendern/private-repo-dependabot-fail
updater | INFO <job_91984925> Checking if org.springframework.boot 2.3.9.RELEASE needs updating
  proxy | 2021/03/05 02:41:14 [014] GET https://plugins.gradle.org:443/m2/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [014] 200 https://plugins.gradle.org:443/m2/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [016] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [016] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
  proxy | 2021/03/05 02:41:15 [018] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [018] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [020] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [020] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [022] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/springframework/boot/org.springframework.boot.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [022] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
updater | INFO <job_91984925> Latest version is 2.4.3
updater | INFO <job_91984925> Pull request already exists for org.springframework.boot with latest version 2.4.3
updater | INFO <job_91984925> Checking if org.testcontainers:postgresql 1.15.1 needs updating
  proxy | 2021/03/05 02:41:15 [024] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/testcontainers/postgresql/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [024] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
  proxy | 2021/03/05 02:41:15 [026] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/testcontainers/postgresql/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [026] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [028] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/testcontainers/postgresql/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [028] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [030] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/testcontainers/postgresql/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [030] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
updater | INFO <job_91984925> Latest version is 
updater | INFO <job_91984925> Requirements to unlock update_not_possible
updater | INFO <job_91984925> Requirements update strategy 
updater | INFO <job_91984925> No update possible for org.testcontainers:postgresql 1.15.1
updater | INFO <job_91984925> Checking if org.projectlombok:lombok 1.18.12 needs updating
  proxy | 2021/03/05 02:41:15 [032] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/projectlombok/lombok/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [032] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
  proxy | 2021/03/05 02:41:15 [034] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/projectlombok/lombok/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [034] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [036] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/projectlombok/lombok/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [036] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
  proxy | 2021/03/05 02:41:15 [038] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/org/projectlombok/lombok/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [038] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
updater | INFO <job_91984925> Latest version is 
updater | INFO <job_91984925> Requirements to unlock update_not_possible
updater | INFO <job_91984925> Requirements update strategy 
updater | INFO <job_91984925> No update possible for org.projectlombok:lombok 1.18.12
updater | INFO <job_91984925> Checking if io.spring.dependency-management 1.0.11.RELEASE needs updating
  proxy | 2021/03/05 02:41:15 [040] GET https://plugins.gradle.org:443/m2/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [040] 200 https://plugins.gradle.org:443/m2/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [042] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [042] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:15 [044] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [044] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
  proxy | 2021/03/05 02:41:15 [046] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:15 [046] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 1.1.1.1:53: no such host
  proxy | 2021/03/05 02:41:16 [048] GET https://nexus.noyoucanthaveaccess.net:443/repository/maven/io/spring/dependency-management/io.spring.dependency-management.gradle.plugin/maven-metadata.xml
  proxy | 2021/03/05 02:41:16 [048] WARN: Cannot read TLS response from mitm'd server dial tcp: lookup nexus.noyoucanthaveaccess.net on 8.8.8.8:53: no such host
updater | INFO <job_91984925> Latest version is 1.0.11.RELEASE
updater | INFO <job_91984925> No update needed for io.spring.dependency-management 1.0.11.RELEASE
updater | INFO <job_91984925> Finished job processing
updater | time="2021-03-05T02:41:16Z" level=info msg="task complete" container_id=job-91984925-updater exit_code=0 job_id=91984925 step=updater
```

**Images of the diff or a link to the PR, issue or logs**

You should be able to fork this repo and reproduce the same issue. 
https://github.com/zendern/private-repo-dependabot-fail

**Other information**
I believe the maven version of dependabot currently will append maven central no matter what you have defined if I'm reading this code correctly. 

https://github.com/dependabot/dependabot-core/blob/f3239dfae253e916b62f9b4b167eeceb6b3ba25a/maven/lib/dependabot/maven/file_parser/repositories_finder.rb#L38-L54

I think the fix might be to add the same thing here. 
https://github.com/dependabot/dependabot-core/blob/f3239dfae253e916b62f9b4b167eeceb6b3ba25a/gradle/lib/dependabot/gradle/file_parser/repositories_finder.rb#L33-L42

Because right now the gradle version only works if you have an artifact repo that is public defined in your setup and in my case I cannot do that. But if dependabot wants to use that and its metadata to tell me I can update that is fine with me. 

I'm happy to open a PR for this please let me know. 

**Current workaround**
For others looking at this issue. If you are on gradle 6.8 or higher you can simply start to use the following configuration and put it in settings.gradle and remove whatever you have in your build.gradle files.

```
pluginManagement {
    repositories {
        maven { url "https://nevergoingtogetit.net/repository/maven/" }
    }
}

dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.PREFER_SETTINGS)
    repositories {
        maven { url "https://nevergoingtogetit.net/repository/maven/" }
    }
}
```

This will configure where plugins are pulled and the new `dependencyResolutionManagement` (https://docs.gradle.org/current/userguide/dependency_management.html#sub:centralized-repository-declaration) allows you to say use this for all projects. The PREFER_SETTINGS flag will simply just ignore anything you put in build.gradle and use what you have in settings.gradle. 

It works b/c the parsing that dependabot does doesn't look at settings.gradle or that new block whatsoever and it just ends up falling back to maven central.

### PR DESCRIPTION
**Pull request by zendern**
Fixes #3236

By following what Maven ecosystem is doing by always adding Maven Central repo to the list.

### CODE REVIEW

File: gradle/lib/dependabot/gradle/file_parser/repositories_finder.rb [hunk 1]
```diff
@@ -36,9 +36,7 @@ def repository_urls
           repository_urls += own_buildfile_repository_urls
           repository_urls = repository_urls.uniq
 
-          return repository_urls unless repository_urls.empty?
-
-          [CENTRAL_REPO_URL]
+          return (repository_urls + [CENTRAL_REPO_URL]).uniq
```
Comment (zendern) [inline gradle/lib/dependabot/gradle/file_parser/repositories_finder.rb:39]: Also `uniq` it here. Not sure if that is necessary but felt like having multiple possibly same repos was the wrong thing to do.

**Review by xlgmokha — COMMENTED** [review]
Thank you for demonstrating a bias for action and for your enthusiasm in getting this resolved. From a quick glance of the changes this looks like a great start.

I have a couple of questions that I'd like to discuss and to collect some additional feedback on such as:

1. Is there a way to detect name squatting collisions?
1. Should falling back to public registries be a default behaviour or an opt-in option?

**Comment by zendern** [comment]
> Thank you for demonstrating a bias for action and for your enthusiasm in getting this resolved. From a quick glance of the changes this looks like a great start.
> 
> I have a couple of questions that I'd like to discuss and to collect some additional feedback on such as:
> 
> 1. Is there a way to detect name squatting collisions?
> 1. Should falling back to public registries be a default behaviour or an opt-in option?
> 

1. Please say more about this, I'll be honest I'm not sure I follow. Are you saying in one public repo (let's say jcenter for example) someone has taken a package name version set and hijacked it but in maven Central it's a valid package?? 

2. Sure I'm happy to make this opt in but should maven follow the same behavior then as well??

**Comment by xlgmokha** [comment]
> 1. Please say more about this, I'll be honest I'm not sure I follow. Are you
> saying in one public repo (let's say jcenter for example) someone has taken
> a package name version set and hijacked it but in maven Central it's a valid
> package??

Let me try to provide an example for clarification:

Private registry: (https://acme.example.org/)
* package name: `acme-logger`
* package version: `0.1.0`

Public registry: (https://repo1.maven.org/maven2/)
* package name: `acme-logger` (malicious package not owned by acme corp)
* package version: `0.1.1`

In the example above an internal company package named "acme-logging" is
hosted and served from a private package registry. At the same time a package
named "acme-logging" is also hosted on a public registry with the same name
but owned by an entity not belonging to acme corp.

In this example, falling back to the public registry would produce
a pull request from Dependabot that suggests installing a malicious
package.

My question is to see how we might be able to identify and prevent
this behaviour.

More info can be found in [Substitution attacks](https://github.blog/2021-02-12-avoiding-npm-substitution-attacks/).

> 2. Sure I'm happy to make this opt in but should maven follow the same
> behavior then as well??

I like consistency. So if this is what we're doing with Maven today
then it would be great to make sure we align the behaviour between
the two.

I raised the question of what the default behaviour
should be to make sure the default is the less risky option.
However, this might also mean adding a way to opt in or choose a specific
behaviour.
If this is the default behaviour for how `mvn` and `gradle` behave then
perhaps this is less of a concern.
I'll need some input from other colleagues on this one.

I might be over thinking this but I'd like to try to prevent introducing
risky packages if possible.

Also, thanks for taking the time to consider the
different security implications. I'd like to see if we can find a way
to cater to the scenario resolved in this PR while also providing safe
defaults for everyone.

What do you think?

**Comment by zendern** [comment]
@xlgmokha Thanks for clarification. That makes sense as far as the attack vector goes. And I agree if we can figure out a way to do that or at least make the behavior opt in that would probably better. 

I did find this while researching substitution/brandjacking attacks. 

https://blog.sonatype.com/malware-removed-from-maven-central


> When a new publisher comes along requesting access to publish to Central, the requirements enforce that you can verify control of either the DNS for the Group ID, or for control over the account/repo for coordinates derived from platforms such as GitHub. As a part of this screening, users are asked to verify their GitHub account before they are assigned a Group ID, such as ‘com.github.codingandcoding’, as is the ID in this case.


Sounds like maven central might make it a little harder to do b/c of the way it validates and requires Group ID. Similar to the article you linked using scopes with npm packages can help protect you there. 

In my case and maybe others packages look something like so 

```
<groupId>com.bigcompany</groupId>
<artifactId>custom-package</artifactId>
```

So you would have to prove you have DNS access to `com.bigcompany` or rights to github. Probably doable still just harder to do. 

Opt-in still feels like maybe the right choice. With the findings below about how the dependency managers work below it might make sense to have maven behave like gradle and not fallback to central unless you opt in. 

---------------

More information on why dependabot probably works the way it does for maven vs gradle. 

I started by creating this branch in my test repo. 

https://github.com/zendern/private-repo-dependabot-fail/tree/default-repo-setup

Where maven and gradle both have no repositories or plugin repositories defined. In the maven case with nothing defined its behavior is to default to maven central. In the gradle case it throws an exception b/c no repositories are defined. 

So I'm guessing the "defaulting/fallback" for maven was to simulate that. But maybe the maven code should only "default/fallback" if no repositories are defined instead of always doing it??

**Comment by xlgmokha** [comment]
> So I'm guessing the "defaulting/fallback" for maven was to simulate that. But maybe the maven code should only "default/fallback" if no repositories are defined instead of always doing it??

I'm checking to see if there is a way to specify opt-in behaviour via the `dependabot.yml` or if there's another way to do this. Meanwhile, I'm also looping in friends from security to make sure I understand the impact of using this as a default. I'll keep you posted as much as I can. Thanks for your patience @zendern.

**Comment by xlgmokha** [comment]
A review from friends in #security suggests that it would be best to avoid falling back to the public registries by default due to recent naming/typo squatting. 

However, our [roadmap](https://github.com/github/roadmap/issues/67) has some exciting features related to better support for private registries. I think this may also cover being able to specify public registries as a fallback mechanism.

I appreciate the effort made to ship this idea. Unfortunately, we wont be able to accept this contribution at this time.

Thank you very much @zendern and keep them coming.

**Comment by zendern** [comment]
@xlgmokha cool. That's fair and I understand. I will close this pr. I assume with the support of private registries there would have to be line of sight from GitHub to that registry??

**Comment by xlgmokha** [comment]
> I assume with the support of private registries there would have to be line of sight from GitHub to that registry??

Yep, that's right. Dependabot would need to be able to resolve the IP address for the private registry host and be able to connect to it with or without credentials. More details to come soon.

**Pull request closed without merging**